segmentstore/server/src/main/java/io/pravega/segmentstore/server/attributes/SegmentAttributeIndex.java

```diff
@@ -0,0 +1,663 @@
+                                        .thenAccept(si -> this.attributeSegment.set(new AttributeSegment(handle, si.getLength()))), this.executor),
+                        ex -> ex instanceof StreamSegmentNotExistsException,
+                        () -> this.storage.create(attributeSegmentName, this.config.getAttributeSegmentRollingPolicy(), timer.getRemaining())
+                                .thenComposeAsync(si -> this.storage.openWrite(attributeSegmentName)
+                                        .thenAccept(handle -> this.attributeSegment.set(new AttributeSegment(handle, si.getLength()))), this.executor)
+                )
+                .thenRun(() -> log.debug("{}: Initialized (Attribute Segment Length = {}).", this.traceObjectId, this.attributeSegment.get().getLength()));
+    }
+
+    /**
+     * Deletes all the Attribute data associated with the given Segment.
+     *
+     * @param segmentMetadata The SegmentMetadata for the Segment whose attribute data should be deleted.
+     * @param storage         A Storage Adapter to execute the deletion on.
+     * @param timeout         Timeout for the operation.
+     * @return A CompletableFuture that, when completed, will indicate that the operation finished successfully.
+     */
+    static CompletableFuture<Void> delete(SegmentMetadata segmentMetadata, Storage storage, Duration timeout) {
+        TimeoutTimer timer = new TimeoutTimer(timeout);
+        String attributeSegmentName = StreamSegmentNameUtils.getAttributeSegmentName(segmentMetadata.getName());
+        return Futures.exceptionallyExpecting(
+                storage.openWrite(attributeSegmentName)
+                       .thenCompose(handle -> storage.delete(handle, timer.getRemaining())),
+                ex -> ex instanceof StreamSegmentNotExistsException,
+                null);
+    }
+
+    //endregion
+
+    //region AttributeIndex implementation
+
+    @Override
+    public CompletableFuture<Void> put(UUID key, Long value, Duration timeout) {
+        return put(Collections.singletonMap(key, value), timeout);
+    }
+
+    @Override
+    public CompletableFuture<Void> put(Map<UUID, Long> values, Duration timeout) {
+        ensureInitialized();
+        Preconditions.checkNotNull(values, "values");
+        if (values.size() == 0) {
+            // Nothing to do.
+            return CompletableFuture.completedFuture(null);
+        } else {
+            AttributeCollection c = new AttributeCollection(values);
+            if (shouldSnapshot()) {
+                // We are overdue for a snapshot. Create one while including the new values. No need to also write them
+                // separately.
+                return createSnapshot(c, false, timeout);
+            } else {
+                // Write the new values separately, as an atomic append.
+                return Futures.toVoid(appendConditionally(() -> CompletableFuture.completedFuture(serialize(c)),
```

Comment:
In the case I'm calling put on a single key-value, I think this is doing an append for that single key-value pair, which is not going to be very efficient.

For consistency, I believe we need this value to be made durable before we can proceed, and Tier 2 writes are not expected to have low latency, so the updates of attributes can hurt performance.

Could you elaborate on what's wrong with my understanding or in general why updating this way is not a problem?

Reply:
This API is only invoked by the `StorageWriter`/`SegmentAggregator` and it already aggregates multiple updates into a single one. It collects all attribute updates for all Appends/Updates it normally aggregates, so a single call to this API will be the result of possibly a lot of individual updates. 

For example, if we have lots of 100-byte appends on the same segment, the SW will wait 30s or until it fills 1MB, thus possibly aggregating 10K such appends together. That means it will include 10K attribute updates, combined into a single call to this API. Also note that multiple updates to the same attribute will be coalesced into a single value (the last one).